docs/reference/content/driver-async/tutorials/gridfs.md

```diff
@@ -0,0 +1,370 @@
+
+```java
+Path outputPath = Paths.get("/tmp/mongodb-tutorial-out.pdf");
+AsynchronousFileChannel streamToDownloadTo = AsynchronousFileChannel.open(outputPath,
+        StandardOpenOption.CREATE_NEW,
+        StandardOpenOption.WRITE,
+        StandardOpenOption.DELETE_ON_CLOSE);
+
+gridFSBucket.downloadToStream(fileId, channelToOutputStream(streamToDownloadTo),
+    new SingleResultCallback<Long>() {
+      @Override
+      public void onResult(final Long result, final Throwable t) {
+          System.out.println("downloaded file sized: " + result);
+      }
+});
+// Note: AsynchronousFileChannel was opened with option delete on close
+streamToDownloadTo.close();
+```
+
+If the `_id` of the file is unknown but you know the filename, then you can pass the filename to the [`downloadToStream`]({{< apiref "com/mongodb/async/client/gridfs/GridFSBucket.html#downloadToStream-java.lang.String-com.mongodb.async.client.gridfs.AsyncOutputStream-com.mongodb.client.gridfs.model.GridFSDownloadOptions-com.mongodb.async.SingleResultCallback-" >}}) method. By default, it will download the latest version of the file. Use the [`GridFSDownloadOptions`]({{< apiref "com/mongodb/client/gridfs/model/GridFSDownloadOptions.html" >}}) to configure which version to download.
+
+```java
+Path outputPath = Paths.get("/tmp/mongodb-tutorial-out.pdf");
+AsynchronousFileChannel streamToDownloadTo = AsynchronousFileChannel.open(outputPath,
+        StandardOpenOption.CREATE_NEW,
+        StandardOpenOption.WRITE,
+        StandardOpenOption.DELETE_ON_CLOSE);
+
+GridFSDownloadOptions downloadOptions = new GridFSDownloadOptions().revision(0);
+
+gridFSBucket.downloadToStream("mongodb-tutorial", channelToOutputStream(streamToDownloadTo), downloadOptions,
+    new SingleResultCallback<Long>() {
+      @Override
+      public void onResult(final Long result, final Throwable t) {
+          System.out.println("downloaded file sized: " + result);
+      }
+});
+// Note: AsynchronousFileChannel was opened with option delete on close
+streamToDownloadTo.close();
+```
+
+### OpenDownloadStream
+
+The [`openDownloadStream`]({{< apiref "com/mongodb/async/client/gridfs/GridFSBucket.html#openDownloadStream-org.bson.types.ObjectId-">}}) method returns a [`GridFSDownloadStream`]({{< apiref "com/mongodb/client/gridfs/GridFSDownloadStream.html">}}) which extends [`AsyncInputStream`]({{< apiref "com/mongodb/async/client/gridfs/AsyncInputStream" >}}).
+
+The following example reads from the `GridFSBucket` via the returned `AsyncInputStream`. For the `fileId` variable, specify an `ObjectId` value returned from the find operation section in this tutorial.
+
+```java
+final ByteBuffer dstByteBuffer = ByteBuffer.allocate(1024 * 1024);
+final GridFSDownloadStream downloadStream = gridFSBucket.openDownloadStream(fileId);
+downloadStream.read(dstByteBuffer, new SingleResultCallback<Integer>() {
+    @Override
+    public void onResult(final Integer result, final Throwable t) {
```

Comment:
CC @rozza 

Same problem here I believe: read may invoke the callback while there is still room is dstByteBuffer, so may have to call read repeatedly to fill it.